README
===============================================================================
README for libicns 0.8.0 - January 19th, 2012
Copyright (C) 2001-2012 Mathew Eis <mathew@eisbox.net>

Other individuals have supported the development of libicns; they are credited
in the source files where their respective contibutions have been made.

===============================================================================
This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA.

===============================================================================
This package include the following libraries and utilities:
1) libicns - A library for the translation of the icns format
2) icns2png - A utility for converting icns files into png images
3) icontainer2icns - A utility for extracting icns files from icontainer packs

You should be able to get the latest version at http://sf.net/projects/icns

===============================================================================
libicns is a library for manipulation of the Mac OS icns resource format, also
known as the IconFamily resource type. It can read and write files from the
Mac OS X icns format, as well as read from Mac OS resource files and macbinary
encoded Mac OS resource forks.

As of release 0.5.9, it can fully manipulate any 128x128 and smaller 32-bit
icons, and has partial support for manipulating 8-bit, 4-bit, and 1-bit icons.

When linked with Jasper, it also has full support for 256x256 and 512x512
icon sizes within the icon family.

As of release 0.8.0, it can read the PNG format introduced with OS X Lion 10.7


Please see DEVNOTES for more information on how to use libicns

===============================================================================
Using icns2png

icns2png -x OmniWeb.icns
Converting OmniWeb.icns to OmniWeb_128x128x32.png...

This will result in a file OmniWeb_128x128x32.png with a 128x128 icon from OmniWeb.icns.

Run icns2png --help for more detailed information.

===============================================================================
Using icontainer2icns

icontainer2icns foo.icontainer

This will end up with a couple of icns files identified by the iContainer ID
and the iContainer name in a (new) folder carrying the icontainer name
(without any suffix)

===============================================================================
NOTICE: icontainer2icns is NOT related to www.iconfactory.com - though it
decodes this library type. They are NOT involved, please don't bother them
with bug reports.

For bug/wishes regarding icontainer2icns mailto: baghira-style@gmx.net

===============================================================================
Understanding the Mac OS X Icons

Mac OS X Icon files come in two formats - icns files, and icns resources
embedded with rsrc files. They may also be stored in a macbinary encoded
rsrc files, or applesingle/appledouble encoded rsrc files

libicns (and icns2png) can decode all five storage formats

The .icns files are much easier to deal with, since they are in a format that
can be easily accesed from Linux/UNIX/Windows.

You can however access the icons from the resource files. OS X Icons like
those from http://interfacelift.com/icons-mac/ are compressed in .sit files.
These files can be uncompressed using Aladdin Stuffit Expander for Linux,
which can be obtained from http://www.stuffit.com/unix/index.html

When unstuffing, be sure to use a command similar to the following:

unstuff --text=auto --macbinary=auto --eol=unix icons.hqx.sit

This will help ensure that the resource forks are uncompressed into
MacBinary files, so that icns2png can read them.

If there are resource forks in the files, when stuffit decompresses them,
it will create two files from the one mac file. For example, if the mac
file was named "Gnu", the resulting decomressed files will be
"Gnu", and "Gnu_1". The "Snowflake_1" file is the resource
fork. This is the file you will want to extract the icon from. The file
"Gnu" does not contain any icon data. If, however, there is no
"Gnu_1", then the file may be in the .icns format, and you will
extract the icon from "Gnu" - resulting in a file "Gnu.png"

It is really fairly simple, once you get the hang of it - you will also
notice that if the icon was in the resource fork, then the other main
file will be etremely small, sometimes even 0k.

The underlying data in both of the files is similar - they both hold the
following icon data (Although most icons to not have a complete set):
16x12 pixels - 1 bit mask - 1, 4, or 8 bit icon
16x16 pixels - 1 or 8 bit mask - 1, 4, 8 or 24 bit icon
32x32 pixels - 1 or 8 bit mask - 1, 4, 8, or 24 bit icon
64x64 pixels - 1 or 8 bit mask - 1, 4, 8, or 24 bit icon
128x128 pixels - 8 bit mask - 24 bit icon


NEWS
Release 0.8.0  (Current)
# Sourceforge SVN rev 170 - current
- libicns version 2.8.2 (per libtool)
- fixed potentially dangerous parsing error
- improved icns2png's understanding of 'icnV' type
- added awareness of 'TOC ' type to libicns/icns2png
- added support for PNG icns elements
- added support for icon variant type
- fix for alignment/bus errors on ARM/Sparc 

Release 0.7.1  (09/03/2009)
# Sourceforge SVN rev 165 - 170
- libicns version 2.7.1 (per libtool)
- 

Release 0.7.0  (07/28/2009)
# Sourceforge SVN rev 163 - 165
- libicns version 2.7.1 (per libtool)
- Fixed licensing concerns

Release 0.6.2  (05/10/2009)
# Sourceforge SVN rev 154 - 162
- libicns version 2.6.1 (per libtool)
- Changed icns_type_t from struct to uint32_t
- Updated all code relevant to the icns_type_t change
- Added a function for creating NULL terminated type strings

Release 0.6.1  (02/05/2009)
# Sourceforge SVN rev 152 - 153
- libicns version 2.6.1 (per libtool)
- Removed debug code from png2icns
- Added CDEF block to jp2 writing code
- Modified icns_set_element_in_family to sort newly inserted elements
- Added icns_type_to_uint32 to libicns internals (for faster type checking)
- Added icns_get_element_order to libicns internals (for sorting elements)
- Minor performance improvements

Release 0.6.0  (01/05/2009)
# Sourceforge SVN rev 149 - 151
- libicns version 2.5.1 (per libtool)
- Fixed bug in with global variable - SF Bug 2139996
- Fixed bug in png2icns; info was not properly updated after transforms

Release 0.5.9  (10/09/2008)
# Sourceforge SVN rev 138 - 148
# libicns version 2.4.1 (per libtool)
- Fixed several small bugs
- Fixed major bug in RLE compressor
- Changed icns2png to save extracted files to current directory by default
- Completed write support for 256x256 & 512x512 via Jasper

Release 0.5.8  (06/12/2008)
# Sourceforge SVN rev 135 - 137
# libicns version 2.3.1 (per libtool) - (NOTE: fix to 2.3.1 at SVN rev 137)
- Fixed licensing issues with FSF address
- Fixed incorrect libicns/libtool versioning

Release 0.5.7  (06/10/2008)
# Sourceforge SVN rev 118 - 134
# libicns version 2.0.1 (per libtool) - (NOTE: should have changed to 2.2.1)
- Updated to use autoconf 2.61, automake 1.10
- Added debain packaging to SVN tree (NOTE: Maintainer use only!)
- Moved libicns/ to src/ for more conventional structure
- Added png2icns utility
- Added confgure check for libjasper and/or libopenjpeg
- Added manpages for png2icns and icontainer2icns
- Prevented libicns from exporting internal-only symbols
- Fixed some compiler errors regarding pointers

Release 0.5.6  (04/10/2008)
# Sourceforge SVN rev 109 - 117
# libicns version 2.0.1 (per libtool) - (NOTE: should have changed to 2.1.1)
- Several write support bugfixes (write support was broken up to this point)
- Minor code clarifications
- Updated icns2png manpage


Release 0.5.5  (03/17/2008)
# Sourceforge SVN rev 102 - 108
# libicns version 2.0.1 (per libtool)
- Converted project to use GNU automake and friends
- Converted all files to UTF-8
- Added multiple batch extraction to icns2png
- Added png output path to icns2png
- Enhanced icns2png for Mac OS X (accesses resource fork now)
- Added HTML API documentation

Release 0.5.2  (03/11/2008)
# Sourceforge SVN rev 93 - 101
- Restructured code tree to seperate icns2png from libicns
- Minor bugfixes

Release 0.5.1  (03/07/2008)
# Sourceforge SVN rev 88 - 92
- Added Jasper JP2 decoder
- Minor bugfixes

Release 0.5.0  (03/05/2008)
# Sourceforge SVN rev 75 - 87
- Added basic API documentation
- Dramatially enhanced handling of endian issues
- Added some minor debug checks
- Changed data types to use stdint.h

Release 0.4.1  (02/28/2008)
# Sourceforge SVN rev 72 - 74
- Fixed libtool issue under Darwin
- Several minor bugfixes

Release 0.3.0  (02/26/2008)
# Sourceforge SVN rev 63 - 71
- Fixed numerous bugs
- Split icns.c into manageable portions
- Added function to retrieve image with mask from icon family
- Modified endian swap routines for easier debugging
- Added routines for accessing unaligned memory
- Made minor code changes for consistency within libicns

Release 0.1.1  (02/20/2008)
# Sourceforge SVN rev 62
- Fixed dynamic library build issues under OS X
- Fixed bugs when uncompressing RLE data in sizes < 128
- Other minor bugfixes
- Added changelog
- Added versioning specifications

Release 0.1.0  (02/18/2008)
- first major release of libicns with write support
- added read support for 1,4,8 bit icons
- removed all possible Apple copyrighted stuff
- added support to modify the icns structure
- added support for 32-bit image manipulation
- added support to do internal rle compression
- added file read/write support using the library directly
- fixed Mac OS X build

Release 0.0.9  (02/14/2008)
- merged in test icns files from original source tree
- added TODO for libicns developers
- finished breakdown of libicns and icns2png
- merged all headers required by libicns into icns.h
- moved jpeg decoding routines into icns.c
- established libicns development naming conventions
- added DEVNOTES

Release 0.0.8  (02/13/2008)
- merged forked source trees from chenxio and baghira
- preliminary split of source into libicns and icns2png

Release 0.0.7  (01/16/2008)
- forked by chenxiao.zhao at Google Code
- added 256x256 and 512x512 support via openjpeg
- added command-line parameters

Release 0.0.6  (08/30/2007)
- forked by baghira at CLI-Apps
- added "NG" icns support
- added icontainter2icns utility

Release 0.0.5 (01/28/2003)
- Fixed bug in resource fork parsing

Release 0.0.4 (01/23/2003)
- Fixed a bug with filenames that was causing segfaults

Release 0.0.3 (12/18/2002)
- Much improved parsing of the icons within resource forks and resource files.
- Much improved MacBinary support.
- Updated README with unstuff information

Release 0.0.2 (01/11/2002)
- Inital public release



src/icns.h
/*
File:       icns.h
Copyright (C) 2001-2008 Mathew Eis <mathew@eisbox.net>
Copyright (C) 2002 Chenxiao Zhao <chenxiao.zhao@gmail.com>

With the exception of the limited portions mentiond, this library
is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published
by the Free Software Foundation; either version 2.1 of the License,
or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
Boston, MA 02110-1301, USA.
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#ifndef _ICNS_H_
#define	_ICNS_H_

/* basic data types */
typedef uint8_t         icns_bool_t;

typedef uint8_t         icns_uint8_t;
typedef int8_t          icns_sint8_t;
typedef uint16_t        icns_uint16_t;
typedef int16_t         icns_sint16_t;
typedef uint32_t        icns_uint32_t;
typedef int32_t         icns_sint32_t;
typedef uint64_t        icns_uint64_t;
typedef int64_t         icns_sint64_t;

typedef uint8_t         icns_byte_t;


/* data header types */
typedef uint32_t        icns_type_t;
typedef int32_t         icns_size_t;

/* icon family and element types */
typedef struct icns_element_t {
  icns_type_t           elementType;    /* 'ICN#', 'icl8', etc... */
  icns_size_t           elementSize;    /* Total size of element  */
  icns_byte_t           elementData[1]; /* icon image data */
} icns_element_t;

typedef struct icns_family_t {
  icns_type_t           resourceType;	/* Always should be 'icns' */
  icns_size_t           resourceSize;	/* Total size of resource  */
  icns_element_t        elements[1];    /* icon elements */
} icns_family_t;

/* icon image data structure */
typedef struct icns_image_t
{
  icns_uint32_t         imageWidth;     // width of image in pixels
  icns_uint32_t         imageHeight;    // height of image in pixels
  icns_uint8_t          imageChannels;  // number of channels in data
  icns_uint16_t         imagePixelDepth;// number of bits-per-pixel
  icns_uint64_t         imageDataSize;  // bytes = width * height * depth / bits-per-pixel
  icns_byte_t           *imageData;     // pointer to base address of uncompressed raw image data
} icns_image_t;

/* used for getting information about various types */
/* not part of the actual icns data format */
typedef struct icns_icon_info_t
{
  icns_type_t           iconType;         // type of icon (or mask)
  icns_bool_t           isImage;          // is this type an image
  icns_bool_t           isMask;           // is this type a mask
  icns_uint32_t         iconWidth;        // width of icon in pixels
  icns_uint32_t         iconHeight;       // height of icon in pixels
  icns_uint8_t          iconChannels;     // number of channels in data
  icns_uint16_t         iconPixelDepth;   // number of bits-per-pixel
  icns_uint16_t         iconBitDepth;     // overall bit depth = iconPixelDepth * iconChannels
  icns_uint64_t         iconRawDataSize;  // uncompressed bytes = width * height * depth / bits-per-pixel
} icns_icon_info_t;

/*  icns element type constants */

#define ICNS_TABLE_OF_CONTENTS        0x544F4320  // "TOC "

#define ICNS_ICON_VERSION             0x69636E56  // "icnV"

#define ICNS_1024x1024_32BIT_ARGB_DATA 0x69633130 // "ic10"

#define ICNS_512x512_32BIT_ARGB_DATA  0x69633039  // "ic09"
#define ICNS_256x256_32BIT_ARGB_DATA  0x69633038  // "ic08"

#define ICNS_128X128_32BIT_DATA       0x69743332  // "it32"
#define ICNS_128X128_8BIT_MASK        0x74386D6B  // "t8mk"

#define ICNS_48x48_1BIT_DATA          0x69636823  // "ich#"
#define ICNS_48x48_4BIT_DATA          0x69636834  // "ich4"
#define ICNS_48x48_8BIT_DATA          0x69636838  // "ich8"
#define ICNS_48x48_32BIT_DATA         0x69683332  // "ih32"
#define ICNS_48x48_1BIT_MASK          0x69636823  // "ich#"
#define ICNS_48x48_8BIT_MASK          0x68386D6B  // "h8mk"

#define ICNS_32x32_1BIT_DATA          0x49434E23  // "ICN#"
#define ICNS_32x32_4BIT_DATA          0x69636C34  // "icl4"
#define ICNS_32x32_8BIT_DATA          0x69636C38  // "icl8"
#define ICNS_32x32_32BIT_DATA         0x696C3332  // "il32"
#define ICNS_32x32_1BIT_MASK          0x49434E23  // "ICN#"
#define ICNS_32x32_8BIT_MASK          0x6C386D6B  // "l8mk"

#define ICNS_16x16_1BIT_DATA          0x69637323  // "ics#"
#define ICNS_16x16_4BIT_DATA          0x69637334  // "ics4"
#define ICNS_16x16_8BIT_DATA          0x69637338  // "ics8"
#define ICNS_16x16_32BIT_DATA         0x69733332  // "is32"
#define ICNS_16x16_1BIT_MASK          0x69637323  // "ics#"
#define ICNS_16x16_8BIT_MASK          0x73386D6B  // "s8mk"

#define ICNS_16x12_1BIT_DATA          0x69636D23  // "icm#"
#define ICNS_16x12_4BIT_DATA          0x69636D34  // "icm4"
#define ICNS_16x12_1BIT_MASK          0x69636D23  // "icm#"
#define ICNS_16x12_8BIT_DATA          0x69636D38  // "icm8"

#define ICNS_32x32_1BIT_ICON          0x49434F4E  // "ICON"

#define ICNS_TILE_VARIANT             0x74696C65  // "tile"
#define ICNS_ROLLOVER_VARIANT         0x6F766572  // "over"
#define ICNS_DROP_VARIANT             0x64726F70  // "drop"
#define ICNS_OPEN_VARIANT             0x6F70656E  // "open"
#define ICNS_OPEN_DROP_VARIANT        0x6F647270  // "odrp"

#define ICNS_NULL_DATA                0x00000000 
#define ICNS_NULL_MASK                0x00000000 

/* icns file / resource type constants */

#define ICNS_FAMILY_TYPE              0x69636E73  // "icns"

#define ICNS_MACBINARY_TYPE           0x6D42494E  // "mBIN"

#define ICNS_NULL_TYPE                0x00000000 

/* icns error return values */

#define	ICNS_STATUS_OK                0

#define	ICNS_STATUS_NULL_PARAM        -1
#define	ICNS_STATUS_NO_MEMORY         -2
#define	ICNS_STATUS_INVALID_DATA      -3

#define	ICNS_STATUS_IO_READ_ERR	      1
#define	ICNS_STATUS_IO_WRITE_ERR      2
#define	ICNS_STATUS_DATA_NOT_FOUND    3
#define	ICNS_STATUS_UNSUPPORTED       4

/* icns function prototypes */
/* NOTE: internal functions are found in icns_internals.h */

// icns_io.c
int icns_write_family_to_file(FILE *dataFile,icns_family_t *iconFamilyIn);
int icns_read_family_from_file(FILE *dataFile,icns_family_t **iconFamilyOut);
int icns_read_family_from_rsrc(FILE *rsrcFile,icns_family_t **iconFamilyOut);
int icns_export_family_data(icns_family_t *iconFamily,icns_size_t *dataSizeOut,icns_byte_t **dataPtrOut);
int icns_import_family_data(icns_size_t dataSize,icns_byte_t *data,icns_family_t **iconFamilyOut);

// icns_family.c
int icns_create_family(icns_family_t **iconFamilyOut);
int icns_count_elements_in_family(icns_family_t *iconFamily, icns_sint32_t *elementTotal);

// icns_element.c
int icns_get_element_from_family(icns_family_t *iconFamily,icns_type_t iconType,icns_element_t **iconElementOut);
int icns_set_element_in_family(icns_family_t **iconFamilyRef,icns_element_t *newIconElement);
int icns_add_element_in_family(icns_family_t **iconFamilyRef,icns_element_t *newIconElement);
int icns_remove_element_in_family(icns_family_t **iconFamilyRef,icns_type_t iconType);
int icns_new_element_from_image(icns_image_t *imageIn,icns_type_t iconType,icns_element_t **iconElementOut);
int icns_new_element_from_mask(icns_image_t *imageIn,icns_type_t iconType,icns_element_t **iconElementOut);
int icns_update_element_with_image(icns_image_t *imageIn,icns_element_t **iconElement);
int icns_update_element_with_mask(icns_image_t *imageIn,icns_element_t **iconElement);

// icns_image.c
int icns_get_image32_with_mask_from_family(icns_family_t *iconFamily,icns_type_t sourceType,icns_image_t *imageOut);
int icns_get_image_from_element(icns_element_t *iconElement,icns_image_t *imageOut);
int icns_get_mask_from_element(icns_element_t *iconElement,icns_image_t *imageOut);
int icns_init_image_for_type(icns_type_t iconType,icns_image_t *imageOut);
int icns_init_image(icns_uint32_t iconWidth,icns_uint32_t iconHeight,icns_uint32_t iconChannels,icns_uint32_t iconPixelDepth,icns_image_t *imageOut);
int icns_free_image(icns_image_t *imageIn);

// icns_rle24.c
int icns_decode_rle24_data(icns_size_t rawDataSize, icns_byte_t *rawDataPtr,icns_size_t expectedPixelCount, icns_size_t *dataSizeOut, icns_byte_t **dataPtrOut);
int icns_encode_rle24_data(icns_size_t dataSizeIn, icns_byte_t *dataPtrIn,icns_size_t *dataSizeOut, icns_byte_t **dataPtrOut);

// icns_jp2.c
int icns_jp2_to_image(icns_size_t dataSize, icns_byte_t *dataPtr, icns_image_t *imageOut);
int icns_image_to_jp2(icns_image_t *image, icns_size_t *dataSizeOut, icns_byte_t **dataPtrOut);

// icns_utils.c
icns_icon_info_t icns_get_image_info_for_type(icns_type_t iconType);
icns_type_t icns_get_mask_type_for_icon_type(icns_type_t);
icns_type_t icns_get_type_from_image_info(icns_icon_info_t iconInfo);
icns_type_t icns_get_type_from_image(icns_image_t iconImage);
icns_type_t icns_get_type_from_mask(icns_image_t iconImage);
icns_bool_t icns_types_equal(icns_type_t typeA,icns_type_t typeB);
icns_bool_t icns_types_not_equal(icns_type_t typeA,icns_type_t typeB);
const char * icns_type_str(icns_type_t type, char *strbuf);
void icns_set_print_errors(icns_bool_t shouldPrint);

#endif
